plat/common/lcpu.c

```diff
@@ -48,3 +165,424 @@ void ukplat_lcpu_halt_to(__nsec until)
+
+int lcpu_fn_enqueue(struct lcpu *lcpu, const struct ukplat_lcpu_func *fn)
+{
+	void (*old_fn)(struct __regs *, const struct ukplat_lcpu_func *);
+
+	UK_ASSERT(fn->fn);
+
+	old_fn = ukarch_load_n(&lcpu->fn.fn);
+
+	/* Check if the slot is empty */
+	if (old_fn != NULL)
+		return -EAGAIN;
+
+	/* It is empty, try to store the function */
+	if (ukarch_compare_exchange_sync(&lcpu->fn.fn, old_fn,
+					 fn->fn) != fn->fn)
+		return -EAGAIN;
+
+	/* We have acquired the slot! Also store the user argument.
+	 * It is safe to do it afterwards, because the RUN IRQ handler will
+	 * only take one function and return afterwards. And we only raise the
+	 * IRQ after finishing setup.
+	 */
+	lcpu->fn.user = fn->user;
+
+	/* Ensure everything is written back when we return and the arch
+	 * support code will raise the IRQ
+	 */
+	wmb();
+
+	return 0;
+}
+
+static void lcpu_fn_dequeue(struct lcpu *this_lcpu, struct ukplat_lcpu_func *fn)
+{
+	*fn = this_lcpu->fn;
+
+	/* Ensure that we have captured the whole function object */
+	rmb();
+
+	UK_ASSERT(fn->fn);
+
+	/* Free the slot. Another function object can be queued afterwards */
+	this_lcpu->fn.fn = NULL;
+}
+
+static int lcpu_ipi_run_handler(void *args __unused, struct __regs *regs)
+{
+	struct lcpu *this_lcpu = lcpu_get_current();
+	struct ukplat_lcpu_func fn;
+
+	lcpu_fn_dequeue(this_lcpu, &fn);
+	fn.fn(regs, &fn);
+
+	/* If we had a transition from BUSY to HALTED in fn, we would
+	 * not reach this code but sit in the error halt loop. We can
+	 * thus safely just decrement without worrying about the HALTED
+	 * state.
+	 */
+	UK_ASSERT(lcpu_state_is_busy(this_lcpu->state));
+	ukarch_dec(&this_lcpu->state);
+
+	return 1;
+}
+
+static int lcpu_ipi_wakeup_handler(void *args __unused,
+				   struct __regs *regs __unused)
+{
+	/* Nothing to do */
+	return 1;
+}
+
+int lcpu_mp_init(void *arg)
+{
+	int rc;
+
+	/* Make sure this is run on the BSP only */
+	UK_ASSERT(lcpu_count == 1);
+	UK_ASSERT(lcpu_get_current() == lcpu_get(0));
+
+	/* Initialize architecture-dependent functionality. This will also do
+	 * CPU discovery and allocation
+	 */
+	rc = lcpu_arch_mp_init(arg);
+	if (unlikely(rc))
+		return rc;
+
+	/* Register the lcpu_run and lcpu_wakeup interrupt handlers */
+	rc = ukplat_irq_register(LCPU_ARCH_IPI_RUN_IRQ,
+				 lcpu_ipi_run_handler, NULL);
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for IPI IRQ %d\n",
+			   LCPU_ARCH_IPI_RUN_IRQ);
+		return rc;
+	}
+
+	rc = ukplat_irq_register(LCPU_ARCH_IPI_WAKEUP_IRQ,
+				 lcpu_ipi_wakeup_handler, NULL);
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for wakeup IRQ %d\n",
+			   LCPU_ARCH_IPI_WAKEUP_IRQ);
+		return rc;
+	}
+
+	return 0;
+}
+
+void __weak __noreturn lcpu_entry_default(struct lcpu *this_lcpu)
+{
+	int rc;
+
+	/* Finish initialization. As there is nothing to return to, we
+	 * just enter halted state if an error occurs.
+	 */
+	rc = lcpu_init(this_lcpu);
+	if (unlikely(rc))
+		lcpu_halt(this_lcpu, rc);
```

Comment:
Do we expect the caller to retry bringing up the core, or halt is halt here? If we need to be able to wake up perhaps we need that be `lcpu_halt_to_irq()`?

Reply:
Actually, at the moment we do not have a state transition back from the `HALTED` state to the `OFFLINE` state. I thought this will probably be also very architecture-specific how to reset a CPU core. We might want to add an extra API call like `ukplat_lcpu_reset` for example in the future. However, we are also missing a public API to recognize that a CPU "crashed" and get the error code. It may also make sense to signal an uk_event or something like that if a CPU crashes.

But this whole topic of error recovery, we do not have thought through enough. For the moment, I think just doing a halt and storing the `rc` is thus enough.